turbo/snapshotsync/freezeblocks/block_reader.go

```diff
@@ -1818,3 +1847,26 @@ func (r *BlockReader) Integrity(ctx context.Context) error {
 	}
 	return nil
 }
+
+func ReadTxNumFuncFromBlockReader(ctx context.Context, r services.FullBlockReader) rawdbv3.ReadTxNumFunc {
+	return func(tx kv.Tx, c kv.Cursor, blockNum uint64) (maxTxNum uint64, ok bool, err error) {
+		maxTxNum, ok, err = rawdbv3.DefaultReadTxNumFunc(tx, c, blockNum)
+		if err != nil {
+			return
+		}
+		if ok {
+			return
+		}
+		hash, err := r.CanonicalHash(ctx, tx, blockNum)
```

Comment:
Let's add method `CanonicalBodyForStorage` to remove `r.CanonicalHash` call here

Reply:
still would use it in case we miss the blockNumber in snapshots as `CanonicalBodies` can be in DB too. also probably not a huge perf overhead.